### ISSUE TITLE
Seige tower target diagonal walls

### ISSUE THREAD

**Issue by GRhin** [Sh0wdown/UnofficialCrusaderPatch#624]
So currently a seige tower can be told to "attack" straight walls, but not diagonal walls. THis makes no sense, and reduces their use alot. Can this be updated so they can be used on any wall configuration? kinda like a ladderman.

**Comment by ByBurton** [comment]
I think laddermen and assassins can't target diagonal walls as well.

**Comment by GRhin** [comment]
Yes they can, I tested them along with the towers.

**Comment by Heroesflorian** [comment]
"Well yes but actually no" :D

Yes, assassins and laddermen can, unlike siege towers, climb diagonal walls _without_ crenellations. 
But, even though the visual indicators differ between siege towers, laddermen and assassins, in fact **_none of the three_** are able to climb diagonal walls _with_ (regular) crenellations. 

I tried to send each of them to "conquer" the crenellated walls here, and none of them moved (although you can see for the assassins that each click did reset their idle-animation if you look closely): 
![yVwP92NawD](https://user-images.githubusercontent.com/12549394/84706952-258cea00-af5e-11ea-9255-52c0912c8c7e.gif)

Moreover, a number of other stuff, including "semi-diagonal walls" with crenellations, any walls with double-layered crenellations, combinations of (uncrenellated) high/low walls, tower-only "walls" as well as moats do also prevent using any of the three units to any effect. 

The particular difference for diagonal walls without crenellations is fairly easy to explain, though: Siege towers are big and bulky, and they need a straight line of 3 wall tiles to attach, and an empty 3x3 tiles space in front of that. A diagonal wall just does not fulfill these requirements (no straight line to attach, and also no free 3x3 tile space in front of any of the tiles). 
But the same (insufficient empty space) does also apply to a gatehouse that is "surrounded" with 1 single tile of wall on each side with a gap of 2 empty tiles between wall and the gate, causing siege towers to not be able to attach to such a gate: 

![grafik](https://user-images.githubusercontent.com/12549394/84704925-e27d4780-af5a-11ea-9487-ac6ff59329c7.png)

And when a siege tower has attached to something, the siege tower will *block* a 3x3 space for building, for as long as it is alive. See below for a comparison of the 1x1 "hitbox" / tile occupation for not-attached siege towers and the 3x3 "hitbox" / tile occupation for attached siege towers:  

![grafik](https://user-images.githubusercontent.com/12549394/84709822-32f8a300-af63-11ea-8e89-74be1293935d.png)
(Lines of regular ground surround the hitbox; tiles of iron ore mark the intended location for building an inn - same relative offset from the siege tower's center in both cases - and yet, next to an attached siege tower, the inn cannot be placed.)

Also, please note that siege towers are generally able to attach to gates (with 3x3 empty space somewhere next to it), even when the gates are closed. In contrast, neither laddermen nor assassins can be ordered to attach to / climb closed gates (assassins will physically attack closed gates instead of climbing them, and laddermen do not even attack gates when they are open): 

![grafik](https://user-images.githubusercontent.com/12549394/84705131-3851ef80-af5b-11ea-92d5-4a475f096a2c.png)

All in all, I would argue siege towers definitely are *not* "kinda like a ladderman", in many ways. 
Siege towers are very sturdy, can be used reliably and by all unit types (friend and foe alike) in order to climb (straight) walls and gates, are expensive, constructed siege engines, and need more space. When destroyed, any troops climbing or standing on a siege tower will survive undamaged, and when not destroyed siege towers can be used as a platform by themselves, even if the attached walls are destroyed. 

![grafik](https://user-images.githubusercontent.com/12549394/84708711-4c005480-af61-11ea-9d9f-cf4cca92b21d.png)

In contrast, laddermen are cheap, fast but vulnerable units that can be used by only a few unit types (own units only) to climb straight and uncrenellated diagonal walls but not gates, need less space and will automatically be picked off by enemy spearmen via instakill. When they die, any troops climbing the ladders will also die instantly, and laddermen cannot be used as platforms (no matter whether the adjacent wall still exists). 

![lrEaklYXt2](https://user-images.githubusercontent.com/12549394/84708002-fa0aff00-af5f-11ea-985d-6e7718aad894.gif)

---

**tl;dr / conclusion:** 
Siege towers are no laddermen, and they should in my opinion not be treated as laddermen or made to be "kinda like laddermen". 

However, all three units - siege towers, laddermen and assassins - are currently very easy to block / deny for any knowledgeable human player or AIV creator, which really is a way more general issue I would say.

**Comment by Monsterfisch** [comment]
I would still argue that the siege tower in its core is a better ladder man and should be treated as such. it has downsides after all which is the space requirements that are very limiting already but he also brings the benefit of being hard to destroy and granting instant access to the top of the wall/gate.

making diagonal walls (for siege towers this could in part be done by making his attached hitbox into a cross instead of a square) be scaleable with and without crenelations seems like a very logical fix for all 3 units fixing.
Arguing that there are other issues with them while correct also doesn't deny that this should be addressed.

Other issues like small wall tall wall and double crenelation scale prevention would be another one.
Those are likely issues that I would tackle in a different way, however, like making it so that the actual merlons actually block vision and actually give additional protection (which would have to be coded in from scratch presumably). 
This way double crenelations would hinder the player more than it helps ergo he would think twice before building double crenelation or at least it would have a "downside" that needs to be considered.

With the lower wall taller wall it could be done that the siege tower gives access to wall tiles in a range instead of just next to it so the lower wall can be jumped if it's just 1 tile thick, this could ofc be countered by making the small wall segment longer, however, fixing assasins to scale a higher wall from a lower one would address that one as well which is one of the things that should be fixed also tbh. It would be sort of realistic that such a wall setup would defend against ladders and siege towers anyways due to their clunkiness. ladders/towers should be more situational anyways to exploit a weakness in the enemy fortification
